docs/Learning-Environment-Design-Agents.md

```diff
@@ -88,81 +100,135 @@ when certain game or simulation events occur, such as in a turn-based game,
-    sensor.AddObservation(ball.transform.GetComponent<Rigidbody>().velocity.z);
+    // Relative position of the ball to the cube (3 floats)
+    sensor.AddObservation(ball.transform.position - gameObject.transform.position);
+    // Velocity of the ball (3 floats)
+    sensor.AddObservation(m_BallRb.velocity);
+    // 8 floats total
 }
 ```
 
-The feature vector must always contain the same number of elements and
-observations must always be in the same position within the list. If the number
-of observed entities in an environment can vary you can pad the feature vector
-with zeros for any missing entities in a specific observation or you can limit
+As an experiment, you can remove the velocity components from
+the observation and retrain the 3DBall agent. While it will learn to balance the
+ball reasonably well, the performance of the agent without using velocity is
+noticeably worse.
+
+The observations passed to `VectorSensor.AddObservation()` must always contain
+the same number of elements must always be in the same order. If the number
+of observed entities in an environment can vary, you can pad the calls
+with zeros for any missing entities in a specific observation, or you can limit
 an agent's observations to a fixed subset. For example, instead of observing
-every enemy agent in an environment, you could only observe the closest five.
+every enemy in an environment, you could only observe the closest five.
 
-When you set up an Agent's `Behavior Parameters` in the Unity Editor, set the
-following properties to use a vector observation:
+Additionally, when you set up an Agent's `Behavior Parameters` in the Unity
+Editor, you must set the **Vector Observations > Space Size**
+to equal the number of floats that are written by `CollectObservations()`.
 
-- **Space Size** — The state size must match the length of your feature vector.
+#### Observable Fields and Properties
+Another approach is to define the relevant observations as fields or properties
+on your Agent class, and annotate them with an ObservableAttribute. For
+example, in the 3DBall example above, the rigid body velocity could be observed
+by adding a property to the Agent:
+```csharp
+using Unity.MLAgents.Sensors.Reflection;
 
-The observation feature vector is a list of floating point numbers, which means
-you must convert any other data types to a float or a list of floats.
+public class Ball3DAgent : Agent {
 
-The `VectorSensor.AddObservation` method provides a number of overloads for
-adding common types of data to your observation vector. You can add Integers and
-booleans directly to the observation vector, as well as some common Unity data
-types such as `Vector2`, `Vector3`, and `Quaternion`.
+    [Observable]
+    public Vector3 RigidBodyVelocity
+    {
+        get { return m_BallRb.velocity;  }
+    }
+}
+```
+In order to enable ObservableAttribute, you must also set "Observable Attribute
+Handling" to "Exclude Inherited" or "Examine All" in the Agent's
```

Comment:
Maybe list all 3 different values and what they do. This would make the paragraph easier to read.

Reply:
Updated.